### PR TITLE
Fiber reset, Bad bits, Hot channels

### PR DESCRIPTION
**Pull request by austinschneider**

### CODE REVIEW

**Comment by slava77** [comment]
Thanks.
Indent-wise it looks much cleaner.
I'll check the actual content later on Monday.

**Comment by austinschneider** [comment]
I tried to clean up Jared's changes to make the code neater in addition to
the indentation issues, but I can already see one place where I might want
to make a style change.

File: emuDCS/PeripheralCore/include/emu/pc/TMB.h [hunk 1]
```diff
@@ -2057,6 +2105,75 @@ class TMB :  public VMEModule, public EMUjtag, public EmuLogger {
...
+  inline int  GetGtxRxAllReset() { return gtx_rx_reset_all_; }
+  inline int  GetReadGtxRxAllReset() { return read_gtx_rx_reset_all_; }
+
+  //Select this GTX for PRBS test input mode
+  inline void SetGtxRxAllPrbsTestEnable(int gtx_rx_prbs_test_enable) { gtx_rx_prbs_test_enable_all_ = gtx_rx_prbs_test_enable; }
+  inline int  GetGtxRxAllPrbsTestEnable() { return gtx_rx_prbs_test_enable_all_; }
+  inline int  GetReadGtxRxAllPrbsTestEnable() { return read_gtx_rx_prbs_test_enable_all_; }
+
+  //Writes all GTX control registers to FPGA
+  void WriteGtxControlRegisters();
```
Comment (slava77) [inline emuDCS/PeripheralCore/include/emu/pc/TMB.h]: this one is extra (second definition) compilation fails

File: emuDCS/PeripheralCore/src/common/TMB.cc [hunk 2]
```diff
@@ -4291,33 +4308,49 @@ void TMB::SetDistripHotChannelMask(int layer,long long int mask) {
...
+  long long m;
+  if (hardware_version_ >= 2){
+	  n = 28;
+	  m = 0x0fffffff;
+  }
+  else {
+    n = 20;
+    m = 0x000fffff;
+  }
+  std::bitset<n> msb_mask;
```
Comment (slava77) [inline emuDCS/PeripheralCore/src/common/TMB.cc]: this doesn't compile

File: emuDCS/PeripheralCore/src/common/TMB.cc [hunk 3]
```diff
@@ -5732,11 +5873,16 @@ void TMB::SetTMBRegisterDefaults() {
...
+    gtx_rx_enable_[i] = gtx_rx_enable_default;
+    gtx_rx_reset_[i] = gtx_rx_reset_default;
+    gtx_rx_prbs_test_enable_[i] = gtx_rx_prbs_test_enable_default;
   }
   //
+  //---------------------------------------------------------------------
+  // 0X15C ADR_V6_CFEB_BADBITS_CTRL: CFEB Bad Bits Control/Status (See Adr 0x122) (extra DCFEB Bad Bits on OTMB)
+  //---------------------------------------------------------------------
+  dcfeb_badbits_reset_ = dcfeb_badbits_reset_default;
+  dcfeb_badbits_block_ = dcfeb_badbits_block_default;
```
Comment (slava77) [inline emuDCS/PeripheralCore/src/common/TMB.cc]: the right hand side dcfeb_badbits_block_default and dcfeb_badbits_reset_default are not defined ==> doesn't compile

**Comment by slava77** [comment]
This looks like an incomplete commit/pull request.
A lot of variables are undefined, the ones where I stopped were badbits related  (dcfeb_badbits_reset_bithi, dcfeb_badbits_reset_bitlo, dcfeb_badbits_reset_default; similar for _block values).

**Comment by austinschneider** [comment]
Pull request was incomplete before. I missed issues seen at compile-time due to being on the wrong branch.
Now successfully compiled on cmlab1 (SLC6, xdaq12).

**Comment by slava77** [comment]
Thanks for fixing quickly.
I can confirm it compiles with the latest code taken from github (as of 3a9ca60)

**Comment by slava77** [comment]
mmcm_qpll_status

Do the numbers of 57007 make sense? It's a 0xDEAF, which looks like a "bad value"
I don't see 0x186 (max address is 0x184) in https://github.com/csc-fw/otmb_fw_docs/blob/master/tmb2005_spec_v5p07.pdf
Where is the current document?

**Comment by slava77** [comment]
the above was for the 904 tmb

```
TMB Firmware version (month/day/year - compile type) : (06/24/2014 - c)...OK...
Firmware Type             : c
Firmware Version Code     : e 
Geographic Address        : 06 
Firmware Revision Code    : 0007 
```

**Comment by slava77** [comment]
Looking at the tmbUtils:
tmbutils_postread

the "read" button  should be renamed, it's too cryptic, "Read GTX Fiber Status" or smth like that

**Comment by slava77** [comment]
When I toggle enable, 
- only "all" actually changes the state on the screen.
- trying to toggle "0" keeps it green (but the text in the printout suggests as if something is happening ... this is probably ok though, because the "all" enabled overrides

```
TMB 0
Toggling fiber enable
Enable before: 1
Enable after: 0
TMBUtils:  TMB 0

```
- if I toggle all disable, then I can actually toggle individual links
  tmbutils_enable1

So, if I enable selectively (0 and 2 are enabled here)
I see that the corresponding TMB status error counts change
status_enable2

I thought that reset would actually reset the count, but it doesn't (well, what do I know).
Austin, have you seen it change in the lab?

**Comment by slava77** [comment]
TMB hard reset clears the error counts

status_afterhardreset

**Comment by austinschneider** [comment]
The 57007 (0xDEAF) is an error code (bad read). Probably the firmware doesn't recognize the command.

**Comment by austinschneider** [comment]
"All" forces an enable, and allows individual toggling when disabled. This is the expected behavior.

The counts should clear, we have seen it in the lab so I don't know what could be happening there. Perhaps it is the wrong firmware version?

**Comment by slava77** [comment]
After running CLCT rx scan more error flags show up, the gtx reset all button doesn't do anything to these counts.

status_afterclctscan

... hard reset clears this

File: emuDCS/PeripheralCore/include/emu/pc/TMB.h [hunk 4]
```diff
@@ -2095,6 +2209,28 @@ class TMB :  public VMEModule, public EMUjtag, public EmuLogger {
...
+  //---------------------------------------------------------------------
+  // 0X15C ADR_V6_CFEB_BADBITS_CTRL: CFEB Bad Bits Control/Status (See Adr 0x122) (extra DCFEB Bad Bits on OTMB)
+  //---------------------------------------------------------------------
+  //!dcfeb_badbits_reset = 1 = Set the reset bit for the "badbits" marker for the last 2 CFEBs
+  void SetDCFEB56BadBitsReset(int dcfeb_badbits_reset);
+  int  GetDCFEB56BadBitsReset();
+  int  GetReadDCFEB56BadBitsReset();
+  //
+  //!dcfeb_badbits_block = 1 = Block channels which have been determined to be "badbits" by the firmware
+  void SetDCFEB56BadBitsBlock(int dcfeb_badbits_block);
```
Comment (slava77) [inline emuDCS/PeripheralCore/include/emu/pc/TMB.h:2228]: this is currently not used, but it seems like it should be just called inside the SetCFEBBadBitsBlock  (the same for the pieces below: reads and writes should go to both the 0..4 and then 5..6 for common cases
... should be on the list of incomplete features, I suppose

**Comment by austinschneider** [comment]
The read button under "Other TMB Utilities" is now less ambiguous.
I also updated the Get/Set methods for the badbits to conditionally include the 2 extra cfebs. It seems best to do that here rather than rely on both methods being called if future code needs to perform those functions.

File: emuDCS/PeripheralCore/src/common/TMB.cc [hunk 5]
```diff
@@ -5246,6 +5413,16 @@ void TMB::DefineTMBConfigurationRegisters_(){
...
+    TMBConfigurationRegister.push_back(hcm501_adr);  //0x16E distrip hot channel mask CFEB 5 layers 0,1
+    TMBConfigurationRegister.push_back(hcm523_adr);  //0x170 distrip hot channel mask CFEB 5 layers 2,3
+    TMBConfigurationRegister.push_back(hcm545_adr);  //0x172 distrip hot channel mask CFEB 5 layers 4,5
+    TMBConfigurationRegister.push_back(hcm601_adr);  //0x174 distrip hot channel mask CFEB 6 layers 0,1
+    TMBConfigurationRegister.push_back(hcm623_adr);  //0x176 distrip hot channel mask CFEB 6 layers 2,3
+    TMBConfigurationRegister.push_back(hcm645_adr);  //0x178 distrip hot channel mask CFEB 6 layers 4,5
+  }
+  //
   // Automatic masking of bad bits
   TMBConfigurationRegister.push_back(cfeb_badbits_ctrl_adr) ;  //0x122
```
Comment (slava77) [inline emuDCS/PeripheralCore/src/common/TMB.cc:5427]: why is dcfeb_badbits_ctrl_adr not here?

File: emuDCS/PeripheralApps/src/common/EmuPeripheralCrateConfig.cc [hunk 6]
```diff
@@ -9511,6 +9649,89 @@ void EmuPeripheralCrateConfig::TMBUtils(xgi::Input * in, xgi::Output * out )
   *out << cgicc::input().set("type","hidden").set("value",buf).set("name","tmb");
   *out << cgicc::form() << std::endl ;
   //
+  std::string TMBFiberReset = toolbox::toString("/%s/TMBFiberReset",
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/EmuPeripheralCrateConfig.cc:9652]: These should be hidden behind if hardware_version>=2

**Pull request merged**